From b30f2a9c60d339faa6568151bef34cc342a3c292 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Wed, 18 Sep 2024 15:14:08 +0200
Subject: [PATCH 1309/1695] drivers: counter: remove dynamic allocation in PM
 handlers

Remove dynamic allocation in stm32_stgen driver PM resume handler
function. The platform should not panic when going into suspend mode
because OP-TEE core heap memory pool is exhausted. Use an allocation
in the stack instead.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 12d3fc463ab0b2c23c460c86c628ba19cd103cde drivers: counter: remove dynamic allocation in PM handlers]
Change-Id: I31ba282902d4ffd8bcc0353d3587a1f2f454c16e
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/405352
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/drivers/counter/stm32_stgen.c | 12 +++---------
 1 file changed, 3 insertions(+), 9 deletions(-)

diff --git a/core/drivers/counter/stm32_stgen.c b/core/drivers/counter/stm32_stgen.c
index c32e46105..24af579f2 100644
--- a/core/drivers/counter/stm32_stgen.c
+++ b/core/drivers/counter/stm32_stgen.c
@@ -112,21 +112,17 @@ static void stm32_stgen_pm_resume(void)
 {
 	unsigned long clock_src_rate = clk_get_rate(stgen_d.tsgen_clock_source);
 	uint64_t counter_val = ((uint64_t)stgen_d.cnt_h << 32) | stgen_d.cnt_l;
-	struct stm32_rtc_calendar *cur_calendar = NULL;
+	struct stm32_rtc_calendar cur_calendar = { };
 	uint64_t nb_pm_count_ticks = 0;
 
-	cur_calendar = calloc(1, sizeof(*cur_calendar));
-	if (!cur_calendar)
-		panic();
-
 	/* STGEN clocks are already restored by the clock driver */
 	io_clrbits32(stgen_d.base + STGENC_CNTCR, BIT(0));
 
 	/* Read the current time from the RTC to update system counter */
-	if (stm32_rtc_get_calendar(cur_calendar))
+	if (stm32_rtc_get_calendar(&cur_calendar))
 		panic("Could not get RTC calendar at resume");
 
-	nb_pm_count_ticks = stm32_rtc_diff_calendar_tick(cur_calendar,
+	nb_pm_count_ticks = stm32_rtc_diff_calendar_tick(&cur_calendar,
 							 stgen_d.calendar,
 							 clock_src_rate);
 
@@ -148,8 +144,6 @@ static void stm32_stgen_pm_resume(void)
 		stm32mp_stgen_smc_config();
 
 	io_setbits32(stgen_d.base + STGENC_CNTCR, BIT(0));
-
-	free(cur_calendar);
 }
 
 static TEE_Result
-- 
2.47.2

